// @stigg/node-server-sdk/dist/utils/apiErrorHandling.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiErrorHandler = void 0;
const core_1 = require("@apollo/client/core");
const lodash_1 = require("lodash");
const sdk_1 = require("@stigg/api-client-js/src/generated/sdk");
const OperationTimeoutError_1 = require("../api/apollo/OperationTimeoutError");
class ExtendedError extends Error {
    constructor(message, errorCode, traceId) {
        super(message);
        this.errorCode = errorCode;
        this.traceId = traceId;
    }
}
function extractErrorData(err) {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    let errorMessage = err.message || 'unknown error';
    let errorCode;
    let traceId;
    const apolloError = err;
    if (!(0, lodash_1.isEmpty)(apolloError.graphQLErrors)) {
        const gqlError = apolloError.graphQLErrors[0];
        errorMessage = gqlError.message;
        errorCode = (_a = gqlError.extensions) === null || _a === void 0 ? void 0 : _a.code;
        traceId = (_b = gqlError.extensions) === null || _b === void 0 ? void 0 : _b.traceId;
    }
    else if (!(0, lodash_1.isEmpty)(apolloError.networkError)) {
        // @ts-expect-error the original error message is on the result object
        const message = ((_d = (_c = apolloError.networkError) === null || _c === void 0 ? void 0 : _c.result) === null || _d === void 0 ? void 0 : _d.errors) && ((_g = (_f = (_e = apolloError.networkError) === null || _e === void 0 ? void 0 : _e.result) === null || _f === void 0 ? void 0 : _f.errors[0]) === null || _g === void 0 ? void 0 : _g.message);
        errorMessage = message || ((_h = apolloError.networkError) === null || _h === void 0 ? void 0 : _h.message) || 'Network error';
    }
    return {
        errorMessage,
        errorCode,
        traceId,
        axiosError: err.isAxiosError ? err : undefined,
    };
}
class ApiErrorHandler {
    constructor(loggerService, eventEmitter) {
        this.loggerService = loggerService;
        this.eventEmitter = eventEmitter;
    }
    async withErrorHandling(func, { errorMessagePrefix, rethrowError }) {
        try {
            const result = await func();
            return result;
        }
        catch (err) {
            const { errorMessage: originalErrorMessage, errorCode, traceId, axiosError } = extractErrorData(err);
            let errorMessage = originalErrorMessage;
            const response = axiosError === null || axiosError === void 0 ? void 0 : axiosError.response;
            if ((response === null || response === void 0 ? void 0 : response.status) === 401 || errorCode === sdk_1.ErrorCode.Unauthenticated) {
                this.eventEmitter.emit('apiKeyInvalid');
                errorMessage = 'Authentication failed. Double check your API key.';
            }
            else {
                const request = axiosError === null || axiosError === void 0 ? void 0 : axiosError.request;
                this.eventEmitter.emit('networkRequestFailed', { errorMessage, errorCode, traceId, request, response });
            }
            if (!rethrowError) {
                this.loggerService.error(`${errorMessagePrefix}: ${err.message}`, err);
                return undefined;
            }
            if (err instanceof core_1.ApolloError && err.networkError instanceof OperationTimeoutError_1.OperationTimeoutError) {
                throw new ExtendedError(`${errorMessagePrefix}: Request timed out`, 'RequestTimeout');
            }
            throw new ExtendedError(`${errorMessagePrefix}: ${errorMessage}`, errorCode, traceId);
        }
    }
}
exports.ApiErrorHandler = ApiErrorHandler;
