// @stigg/node-server-sdk/dist/utils/CacheMapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const dateUtils_1 = require("./dateUtils");
class CacheMapper {
    mapCachedEntitlement(entitlement) {
        var _a, _b, _c, _d;
        let feature = undefined;
        if (entitlement.feature) {
            feature = {
                id: (_a = entitlement.feature) === null || _a === void 0 ? void 0 : _a.refId,
                featureType: entitlement.feature.featureType,
                meterType: entitlement.feature.meterType || undefined,
                units: entitlement.feature.featureUnits || '',
                unitsPlural: entitlement.feature.featureUnitsPlural || '',
            };
        }
        return {
            calculatedEntitlement: {
                feature,
                resourceId: entitlement.resourceId || undefined,
                usageLimit: entitlement.usageLimit,
                hasUnlimitedUsage: entitlement.hasUnlimitedUsage,
                hasSoftLimit: !!entitlement.hasSoftLimit,
                accessDeniedReason: entitlement.accessDeniedReason || undefined,
                enumValues: entitlement.enumValues || [],
            },
            featureUsage: {
                resetPeriod: entitlement.resetPeriod,
                resetPeriodConfiguration: entitlement.resetPeriodConfiguration,
                currentUsage: entitlement.currentUsage || 0,
                usagePeriodAnchor: (_b = (0, dateUtils_1.mapDate)(entitlement.usagePeriodAnchor)) === null || _b === void 0 ? void 0 : _b.getTime(),
                usagePeriodStart: (_c = (0, dateUtils_1.mapDate)(entitlement.usagePeriodStart)) === null || _c === void 0 ? void 0 : _c.getTime(),
                usagePeriodEnd: (_d = (0, dateUtils_1.mapDate)(entitlement.usagePeriodEnd)) === null || _d === void 0 ? void 0 : _d.getTime(),
            },
        };
    }
    mapCachedEntitlements(updatedEntitlements) {
        return new Map(updatedEntitlements.map((entitlement) => {
            var _a;
            return [
                ((_a = entitlement.feature) === null || _a === void 0 ? void 0 : _a.refId) || '',
                this.mapCachedEntitlement(entitlement),
            ];
        }));
    }
}
exports.default = CacheMapper;
