// @stigg/node-server-sdk/dist/services/EdgeApiClient.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EdgeApiClient = void 0;
const axios_1 = __importDefault(require("axios"));
const core_1 = require("@apollo/client/core");
const requestHeaders_1 = require("../utils/requestHeaders");
const axios_retry_1 = __importDefault(require("axios-retry"));
const http = __importStar(require("http"));
const https = __importStar(require("https"));
const lodash_1 = require("lodash");
const REQUEST_RETRY_COUNT = 3;
class EdgeApiClient {
    constructor(config, loggerService) {
        this.config = config;
        this.loggerService = loggerService;
        const { apiKey, baseEdgeUri } = config;
        this.httpClient = this.initHttpClient(baseEdgeUri, apiKey);
    }
    static create(config, loggerService) {
        return config.enableEdge ? new EdgeApiClient(config, loggerService) : null;
    }
    getPaywall(productId, billingCountryCode, fetchAllCountriesPrices, includeHiddenPlans) {
        const prefix = productId ? `/p/${productId}` : '';
        const fetchAllCountriesPricesQueryParam = `fetchAllCountriesPrices=${fetchAllCountriesPrices}`;
        let url = `/v1${prefix}/paywall.json?${fetchAllCountriesPricesQueryParam}`;
        if (billingCountryCode) {
            url += `&billingCountryCode=${billingCountryCode}`;
        }
        if (includeHiddenPlans) {
            url += `&includeHiddenPlans=${includeHiddenPlans}`;
        }
        return this.get(url);
    }
    getEntitlements(customerId, resourceId) {
        const url = `/v1/c/${customerId}/entitlements.json${resourceId ? `?resourceId=${resourceId}` : ''}`;
        return this.get(url, {
            timeout: this.config.entitlementsTimeout,
        });
    }
    getActiveSubscriptionsList(customerId, resourceId) {
        const queryString = new URLSearchParams();
        if ((0, lodash_1.isString)(resourceId)) {
            queryString.set('resourceId', resourceId);
        }
        else if ((0, lodash_1.isArray)(resourceId)) {
            resourceId.forEach((x) => queryString.append('resourceIds', x));
        }
        const url = `/v1/c/${customerId}/subscriptions.json?${queryString.toString()}`;
        return this.get(url);
    }
    getSdkConfiguration() {
        return this.get(`/v1/config/server-sdk-configuration.json`);
    }
    async get(url, options) {
        const response = await this.httpClient.get(url, options);
        const { data } = response;
        if (!(0, lodash_1.isEmpty)(data.errors)) {
            throw new core_1.ApolloError({ graphQLErrors: data.errors });
        }
        return data;
    }
    initHttpClient(baseEdgeUri, apiKey) {
        const axiosInstance = axios_1.default.create({
            baseURL: baseEdgeUri,
            headers: Object.assign(Object.assign({}, (0, requestHeaders_1.buildRequestHeaders)(apiKey, this.loggerService.getInstanceId())), { 'X-API-VERSION': '1' }),
            timeout: this.config.timeout,
            httpAgent: new http.Agent({ keepAlive: true }),
            httpsAgent: new https.Agent({ keepAlive: true }),
        });
        this.enableRetries(axiosInstance);
        return axiosInstance;
    }
    /**
     * Public for mocking in tests
     */
    enableRetries(axiosInstance) {
        (0, axios_retry_1.default)(axiosInstance, {
            retries: REQUEST_RETRY_COUNT,
            retryCondition: (error) => this.canRetryRequest(error),
            retryDelay: (retryNumber, error) => axios_retry_1.default.exponentialDelay(retryNumber, error, 2000),
            shouldResetTimeout: false,
            onRetry: async (retryCount, error, requestConfig) => this.loggerService.log(`Retrying request ${requestConfig.url} - attempt ${retryCount}`, {
                previousError: error.message,
            }),
        });
    }
    canRetryRequest(error) {
        const isTimeoutError = error.code === 'ECONNABORTED' && error.message.includes('timeout');
        return isTimeoutError || axios_retry_1.default.isNetworkOrIdempotentRequestError(error);
    }
}
exports.EdgeApiClient = EdgeApiClient;
