// @stigg/node-server-sdk/dist/services/initializationStateTracker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InitializationStateTracker = exports.InitializationState = void 0;
var InitializationState;
(function (InitializationState) {
    InitializationState["IN_PROGRESS"] = "IN_PROGRESS";
    InitializationState["SUCCESS"] = "READY";
    InitializationState["FAILED"] = "FAILED";
})(InitializationState = exports.InitializationState || (exports.InitializationState = {}));
class InitializationStateTracker {
    constructor(eventEmitter) {
        this.eventEmitter = eventEmitter;
        this.initializationPromise = null;
        this.initializationStatus = InitializationState.IN_PROGRESS;
        this.initFailureError = null;
    }
    getInitializationPromise() {
        if (this.initializationPromise) {
            return this.initializationPromise;
        }
        if (this.initializationStatus === InitializationState.SUCCESS) {
            return Promise.resolve();
        }
        if (this.initializationStatus === InitializationState.FAILED) {
            return Promise.reject(this.initFailureError);
        }
        this.initializationPromise = new Promise((resolve, reject) => {
            const onSuccess = () => {
                resolve();
            };
            const onFailure = (err) => {
                reject(err);
            };
            this.eventEmitter.once('initializeSuccessfully', onSuccess);
            this.eventEmitter.once('initializeFailed', onFailure);
        });
        return this.initializationPromise;
    }
    signalSuccessInit() {
        this.initializationStatus = InitializationState.SUCCESS;
        this.eventEmitter.emit('initializeSuccessfully');
    }
    signalFailedInit(err) {
        this.initializationStatus = InitializationState.FAILED;
        this.initFailureError = err;
        this.eventEmitter.emit('initializeFailed', err);
    }
}
exports.InitializationStateTracker = InitializationStateTracker;
