// @stigg/node-server-sdk/dist/services/redisEntitlementsService.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedisEntitlementsService = void 0;
const entitlementsService_1 = require("./entitlementsService");
const ReuseOngoingExecution_1 = require("../utils/decorators/ReuseOngoingExecution");
const lodash_1 = require("lodash");
class RedisEntitlementsService extends entitlementsService_1.EntitlementsService {
    constructor(cacheService, graphqlClient, loggerService, edgeApiClient, entitlementsTimeout) {
        super(graphqlClient, loggerService, edgeApiClient, cacheService, entitlementsTimeout);
        this.cacheService = cacheService;
        this.graphqlClient = graphqlClient;
        this.loggerService = loggerService;
        this.isInitialized = true; // redis is always considered as initialized
    }
    async refetchEntityEntitlementsAfterCacheMiss(params) {
        if (!this.cacheService.isClientConnected()) {
            throw new Error('Redis is not connected. Cannot refetch entitlements.');
        }
        const { distributedRefetchEntitlementsService } = this.cacheService;
        if (!distributedRefetchEntitlementsService) {
            return this.refetchEntitlements(params);
        }
        const { customerId, resourceId } = params;
        const reloadFromApi = () => this.refetchEntitlements(params);
        const loadFromCache = () => this.cacheService.getCustomerEntitlements(customerId, resourceId);
        return distributedRefetchEntitlementsService.executeOnceAcrossMultipleProcesses({
            customerId,
            resourceId,
            reloadFromApi,
            loadFromCache,
        });
    }
    clearCache() {
        // Clearing persisted cache is not supported
        return Promise.resolve();
    }
    init() {
        // Real time updated via sdk is not supported with persisted cache
        return;
    }
    cleanup() {
        return this.cacheService.cleanup();
    }
}
__decorate([
    (0, ReuseOngoingExecution_1.ReuseOngoingExecution)(({ customerId, resourceId }) => (0, lodash_1.compact)([customerId, resourceId]).join(':'))
], RedisEntitlementsService.prototype, "refetchEntityEntitlementsAfterCacheMiss", null);
exports.RedisEntitlementsService = RedisEntitlementsService;
