// @stigg/node-server-sdk/dist/services/entitlementsService.utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.entitlementsResponseMapper = exports.doesCustomerExist = exports.DATE_IN_FAR_PAST = void 0;
const lodash_1 = require("lodash");
exports.DATE_IN_FAR_PAST = new Date(0);
// The getEntitlements API returns empty list for both:
//   * customer not found
//   * customer exists but doesn't have any subscriptions.
// So the SDK side can't differentiate between them, and as a
// compromise we consider empty entitlements as customer not found.
const doesCustomerExist = (entitlements) => !(0, lodash_1.isEmpty)(entitlements);
exports.doesCustomerExist = doesCustomerExist;
exports.entitlementsResponseMapper = {
    cacheHit: (entitlements) => ({
        cacheMiss: false,
        entitlements,
        customerExists: (0, exports.doesCustomerExist)(entitlements),
        globalCustomerMissing: false,
    }),
    cacheMiss: (globalCustomerMissing = false) => ({
        cacheMiss: true,
        entitlements: null,
        customerExists: false,
        globalCustomerMissing,
    }),
};
