// @stigg/node-server-sdk/dist/services/inMemoryEntitlementsService.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InMemoryEntitlementsService = void 0;
const sdk_1 = require("@stigg/api-client-js/src/generated/sdk");
const entitlementsService_1 = require("./entitlementsService");
const ReuseOngoingExecution_1 = require("../utils/decorators/ReuseOngoingExecution");
const lodash_1 = require("lodash");
const WS_ERROR_RECONNECT_INTERVAL_MS = 5 * 1000;
class InMemoryEntitlementsService extends entitlementsService_1.EntitlementsService {
    constructor(cacheService, client, eventEmitter, loggerService, edgeApiClient, entitlementsTimeout) {
        super(client, loggerService, edgeApiClient, cacheService, entitlementsTimeout);
        this.cacheService = cacheService;
        this.eventEmitter = eventEmitter;
        this.entitlementsSubscription = null;
        this.usageSubscription = null;
        this.packagePublished = null;
    }
    init() {
        this.loggerService.log('Subscribing to real-time updates');
        this.cleanup();
        this.entitlementsSubscription = this.entitlementsApi.subscribeEntitlementsUpdated().subscribe((value) => this.onEntitlementsUpdated(value), (err) => this.websocketErrorHandling(err));
        this.usageSubscription = this.entitlementsApi.subscribeUsageUpdated().subscribe((value) => this.onUsageUpdated(value), (err) => this.websocketErrorHandling(err));
        this.packagePublished = this.entitlementsApi.subscribePackagePublished().subscribe((value) => this.onPackagePublished(value), (err) => this.websocketErrorHandling(err));
        this.isInitialized = true;
    }
    websocketErrorHandling(err) {
        const isErrorRecoverable = this.isErrorRecoverable(err);
        const actionDescription = isErrorRecoverable ? 'retrying to connect' : 'giving up permanently';
        const errorReason = err.message || err.reason;
        const logMessage = `Received error while handling websocket events. Error: ${errorReason} - ${actionDescription}`;
        const isThrottleError = errorReason.includes('ThrottlerException');
        if (isThrottleError) {
            this.loggerService.log(logMessage);
        }
        else {
            this.loggerService.error(logMessage);
        }
        if (isErrorRecoverable) {
            this.cleanup();
            setTimeout(() => this.init(), WS_ERROR_RECONNECT_INTERVAL_MS);
        }
    }
    cleanup() {
        var _a, _b, _c;
        (_a = this.entitlementsSubscription) === null || _a === void 0 ? void 0 : _a.unsubscribe();
        this.entitlementsSubscription = null;
        (_b = this.usageSubscription) === null || _b === void 0 ? void 0 : _b.unsubscribe();
        this.usageSubscription = null;
        (_c = this.packagePublished) === null || _c === void 0 ? void 0 : _c.unsubscribe();
        this.packagePublished = null;
        this.cacheService.clearCache();
        this.isInitialized = false;
    }
    onEntitlementsUpdated(result) {
        if (result.errors || !result.data) {
            this.loggerService.error(`Failed to handle entitlements updated event. Errors: ${result.errors}`, {
                errors: JSON.stringify(result.errors),
            });
            return;
        }
        const entitlementsUpdated = result.data.entitlementsUpdated;
        this.loggerService.debug(`Entitlements updated event arrived.`, {
            customerId: entitlementsUpdated.customerId,
            entitlements: entitlementsUpdated.entitlements,
        });
        const entitlements = this.cacheMapper.mapCachedEntitlements(entitlementsUpdated.entitlements);
        this.cacheService.setCustomer(entitlementsUpdated.customerId, entitlements, entitlementsUpdated.resourceId || undefined);
        this.eventEmitter.emit('entitlementsUpdated', entitlementsUpdated);
    }
    onUsageUpdated(result) {
        if (result.errors || !result.data) {
            this.loggerService.error(`Failed to handle usage updated event. Errors: ${result.errors}`, {
                errors: JSON.stringify(result.errors),
            });
            return;
        }
        const { usage: { customerId, featureId, resourceId, currentUsage, usagePeriodStart, usagePeriodEnd }, } = result.data.usageUpdated;
        this.loggerService.debug(`Usage updated event arrived.`, {
            customerId: customerId,
            featureId: featureId,
            resourceId: resourceId,
            currentUsage: currentUsage,
            usagePeriodStart: usagePeriodStart,
            usagePeriodEnd: usagePeriodEnd,
        });
        const wasSet = this.cacheService.updateFeatureUsage({
            timestamp: new Date(),
            customerId,
            resourceId: resourceId || undefined,
            featureId,
            currentUsage,
            usagePeriodStart: usagePeriodStart || undefined,
            usagePeriodEnd: usagePeriodEnd || undefined,
        });
        if (wasSet) {
            this.eventEmitter.emit('usageUpdated', result.data.usageUpdated);
        }
    }
    isErrorRecoverable(err) {
        const isInvalidAPIKeyError = err.reason === 'Invalid API key';
        return !isInvalidAPIKeyError;
    }
    clearCache() {
        return this.cacheService.clearCache();
    }
    onPackagePublished(result) {
        var _a, _b;
        if (result.errors || !result.data) {
            this.loggerService.error(`Failed to handle package published event. Errors: ${result.errors}`, {
                errors: JSON.stringify(result.errors),
            });
            return;
        }
        this.loggerService.debug(`Package published event arrived.`, {
            data: (_a = result.data) === null || _a === void 0 ? void 0 : _a.packagePublished,
        });
        if (((_b = result.data) === null || _b === void 0 ? void 0 : _b.packagePublished.migrationType) === sdk_1.PublishMigrationType.AllCustomers) {
            this.loggerService.log('Clearing cache due to package published event');
            this.clearCache();
        }
        this.eventEmitter.emit('packagePublished', result.data.packagePublished);
    }
    async refetchEntityEntitlementsAfterCacheMiss(params) {
        return this.refetchEntitlements(params);
    }
}
__decorate([
    (0, ReuseOngoingExecution_1.ReuseOngoingExecution)(({ customerId, resourceId }) => (0, lodash_1.compact)([customerId, resourceId]).join(':'))
], InMemoryEntitlementsService.prototype, "refetchEntityEntitlementsAfterCacheMiss", null);
exports.InMemoryEntitlementsService = InMemoryEntitlementsService;
