// @stigg/node-server-sdk/dist/services/loggerService.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoggerService = exports.LogLevel = void 0;
const includes_1 = __importDefault(require("lodash/includes"));
const STIGG_LOG_PREFIX = '[@stigg/node-server-sdk]';
var LogLevel;
(function (LogLevel) {
    LogLevel["ALL"] = "ALL";
    LogLevel["NONE"] = "NONE";
    LogLevel["LOG"] = "LOG";
    LogLevel["ERROR"] = "ERROR";
})(LogLevel = exports.LogLevel || (exports.LogLevel = {}));
class LoggerService {
    constructor(logger, logLevel) {
        this.logger = logger;
        this.logLevel = logLevel;
    }
    error(message, error) {
        if (this.logLevel !== LogLevel.NONE) {
            const aMessage = this.prependPrefix(message);
            if (error) {
                this.logger.error(aMessage, error);
            }
            else {
                this.logger.error(aMessage);
            }
        }
    }
    log(message, object) {
        if (!(0, includes_1.default)([LogLevel.ERROR, LogLevel.NONE], this.logLevel)) {
            const aMessage = this.prependPrefix(message);
            if (object) {
                this.logger.log(aMessage, object);
            }
            else {
                this.logger.log(aMessage);
            }
        }
    }
    debug(message, object) {
        if ((0, includes_1.default)([LogLevel.LOG, LogLevel.ALL], this.logLevel)) {
            const aMessage = this.prependPrefix(message);
            if (object) {
                this.logger.debug(aMessage, object);
            }
            else {
                this.logger.debug(aMessage);
            }
        }
    }
    prependPrefix(message) {
        return `${STIGG_LOG_PREFIX} [iid:${this.instanceId || '-'}] ${message}`;
    }
    setInstanceId(instanceId) {
        this.instanceId = instanceId;
    }
    getInstanceId() {
        return this.instanceId;
    }
}
exports.LoggerService = LoggerService;
