// @stigg/node-server-sdk/dist/services/cache/inMemoryCacheService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InMemoryCacheService = exports.getApproxSizeBytes = void 0;
const cacheKeysHelpers_1 = require("../../utils/cacheKeysHelpers");
const lru_cache_1 = require("lru-cache");
const entitlementsService_utils_1 = require("../entitlementsService.utils");
const textEncoder = new TextEncoder();
function getApproxSizeBytes(obj) {
    return textEncoder.encode(typeof obj === 'string' ? obj : JSON.stringify(obj)).length;
}
exports.getApproxSizeBytes = getApproxSizeBytes;
class InMemoryCacheService {
    constructor(maxSizeBytes, instrumentation) {
        this.instrumentation = instrumentation;
        this.clearCache = () => {
            this.entitlements.clear();
        };
        const maxSize = maxSizeBytes;
        this.entitlements = new lru_cache_1.LRUCache({
            maxSize,
            sizeCalculation: (value) => getApproxSizeBytes(value),
        });
    }
    waitForInitialization() {
        return Promise.resolve();
    }
    setCustomer(customerId, customerEntitlements, resourceId) {
        const cachedCustomerKey = (0, cacheKeysHelpers_1.buildCachedCustomerKey)(customerId, resourceId);
        this.entitlements.set(cachedCustomerKey, customerEntitlements);
    }
    getCustomerEntitlements(customerId, resourceId) {
        const cachedCustomerKey = (0, cacheKeysHelpers_1.buildCachedCustomerKey)(customerId, resourceId);
        const entitlements = this.entitlements.get(cachedCustomerKey) || null;
        if (entitlements) {
            this.instrumentation.trackHit({ customerId, resourceId });
            return entitlementsService_utils_1.entitlementsResponseMapper.cacheHit(entitlements);
        }
        else {
            this.instrumentation.trackMiss({ customerId, resourceId });
            return entitlementsService_utils_1.entitlementsResponseMapper.cacheMiss();
        }
    }
    updateFeatureUsage({ customerId, featureId, currentUsage, resourceId, usagePeriodStart, usagePeriodEnd, }) {
        const cachedCustomerKey = (0, cacheKeysHelpers_1.buildCachedCustomerKey)(customerId, resourceId);
        const customerEntitlements = this.entitlements.get(cachedCustomerKey);
        const existingFeatureEntitlement = customerEntitlements === null || customerEntitlements === void 0 ? void 0 : customerEntitlements.get(featureId);
        if (!customerEntitlements || !existingFeatureEntitlement) {
            // do nothing in this case, we've never checked this for customer entitlements yet
            return false;
        }
        const { featureUsage: existingFeatureUsage, calculatedEntitlement } = existingFeatureEntitlement;
        customerEntitlements.set(featureId, {
            calculatedEntitlement,
            featureUsage: Object.assign(Object.assign({}, existingFeatureUsage), { currentUsage, usagePeriodStart, usagePeriodEnd }),
        });
        return true;
    }
    getCustomerEntitlement(featureId, customerId, resourceId) {
        const { entitlements, customerExists, cacheMiss, globalCustomerMissing } = this.getCustomerEntitlements(customerId, resourceId);
        const entitlement = !cacheMiss ? (entitlements === null || entitlements === void 0 ? void 0 : entitlements.get(featureId)) || null : null;
        return { cacheMiss, customerExists, entitlement, globalCustomerMissing };
    }
}
exports.InMemoryCacheService = InMemoryCacheService;
