// @stigg/node-server-sdk/dist/api/entitlements/entitlementsApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const sdk_1 = require("@stigg/api-client-js/src/generated/sdk");
class EntitlementsApi {
    constructor(client, edgeApiClient, loggerService) {
        this.client = client;
        this.edgeApiClient = edgeApiClient;
        this.loggerService = loggerService;
    }
    async getEntitlements(customerId, resourceId, skipEdge = false, timeout) {
        // fetching entitlements from edge-api instead of graphql-api
        if (this.edgeApiClient && !skipEdge) {
            this.loggerService.debug(`Fetching entitlements from edge`, { customerId, resourceId });
            return this.edgeApiClient.getEntitlements(customerId, resourceId);
        }
        this.loggerService.debug(`Fetching entitlements from api`, { customerId });
        return this.client.query({
            query: sdk_1.GetEntitlementsDocument,
            variables: { query: { customerId, resourceId } },
            fetchPolicy: 'no-cache',
            context: {
                timeout,
            },
        });
    }
    reportUsage(data) {
        return this.client.mutate({
            mutation: sdk_1.ReportUsageDocument,
            variables: { input: data },
            fetchPolicy: 'no-cache',
        });
    }
    reportUsageBulk(input) {
        return this.client.mutate({
            mutation: sdk_1.ReportUsageBulkDocument,
            variables: { input },
            fetchPolicy: 'no-cache',
        });
    }
    reportEvent(data) {
        return this.client.mutate({
            mutation: sdk_1.ReportEventDocument,
            variables: { input: { usageEvents: data } },
            fetchPolicy: 'no-cache',
        });
    }
    subscribeEntitlementsUpdated() {
        return this.client.subscribe({
            query: sdk_1.OnEntitlementsUpdatedDocument,
        });
    }
    subscribeUsageUpdated() {
        return this.client.subscribe({
            query: sdk_1.OnUsageUpdatedDocument,
        });
    }
    subscribePackagePublished() {
        return this.client.subscribe({
            query: sdk_1.OnPackagePublishedDocument,
        });
    }
}
exports.default = EntitlementsApi;
