// @stigg/node-server-sdk/dist/configuration.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateConfiguration = exports.getConfiguration = exports.DEFAULT_ENTITLEMENTS_TIMEOUT = exports.DEFAULT_TIMEOUT = exports.SDK_NAME = exports.SDK_SOURCE_TYPE = void 0;
const sdk_1 = require("@stigg/api-client-js/src/generated/sdk");
const loggerService_1 = require("./services/loggerService");
const lodash_1 = require("lodash");
const v8_1 = __importDefault(require("v8"));
exports.SDK_SOURCE_TYPE = sdk_1.SourceType.NodeServerSdk;
exports.SDK_NAME = 'node-server-sdk';
exports.DEFAULT_TIMEOUT = 60 * 1000;
exports.DEFAULT_ENTITLEMENTS_TIMEOUT = 10 * 1000;
const LOCALHOST_IPS = ['localhost', '127.0.0.1', '::1'];
function getDefaultMemoryCacheMaxSizeBytes() {
    return v8_1.default.getHeapStatistics().total_available_size * 0.75;
}
const defaults = () => {
    const baseUri = 'https://api.stigg.io';
    return {
        baseUri,
        wsUrl: getWsUrl(baseUri),
        baseEdgeUri: getBaseEdgeApiUri(baseUri),
        logConfiguration: {
            logger: console,
            logLevel: loggerService_1.LogLevel.ERROR,
        },
        timeout: exports.DEFAULT_TIMEOUT,
        entitlementsTimeout: exports.DEFAULT_ENTITLEMENTS_TIMEOUT,
        realtimeUpdatesEnabled: true,
        enableEdge: true,
        memoryCacheMaxSizeBytes: getDefaultMemoryCacheMaxSizeBytes(),
        enableRemoteConfig: true,
    };
};
const defaultRedisOptions = () => ({
    commandTimeout: 60000,
    retryStrategy: (times) => {
        const baseDelay = Math.min(2 ** times * 100, 8000);
        const jitter = Math.floor(Math.random() * 500);
        return baseDelay + jitter;
    },
});
function getConfiguration(configuration) {
    const defaultConfig = defaults();
    const wsUrl = configuration.wsUrl
        ? configuration.wsUrl
        : configuration.baseUri
            ? getWsUrl(configuration.baseUri)
            : defaultConfig.wsUrl;
    const baseEdgeUri = configuration.baseEdgeUri
        ? configuration.baseEdgeUri
        : configuration.baseUri
            ? getBaseEdgeApiUri(configuration.baseUri)
            : defaultConfig.baseEdgeUri;
    const redis = configuration.redis ? (0, lodash_1.merge)(defaultRedisOptions(), configuration.redis) : undefined;
    return (0, lodash_1.merge)(defaultConfig, configuration, { wsUrl, baseEdgeUri, redis });
}
exports.getConfiguration = getConfiguration;
function getWsUrl(apiUrl) {
    const url = new URL(apiUrl);
    if (url.protocol === 'http:') {
        url.protocol = 'ws:';
    }
    else if (url.protocol === 'https:') {
        url.protocol = 'wss:';
    }
    // Remove last char since URL.toString() adds a redundant backslash
    return url.toString().slice(0, -1);
}
function getBaseEdgeApiUri(apiUrl) {
    const url = new URL(apiUrl);
    if (LOCALHOST_IPS.includes(url.hostname)) {
        url.port = '3060';
    }
    else {
        url.hostname = `edge.${url.hostname}`;
    }
    // Remove last char since URL.toString() adds a redundant backslash
    return url.toString().slice(0, -1);
}
/**
 * We are validating here configuration that is required in order to initiate a new client, otherwise we return error
 */
function validateConfiguration(configuration, loggerService) {
    if (!configuration.apiKey) {
        const error = 'You must configure the client with an API key';
        loggerService.error(error);
        throw new Error(error);
    }
}
exports.validateConfiguration = validateConfiguration;
