// @stigg/node-server-sdk/dist/models.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomerModel = exports.ENUM_DEFAULT_FALLBACK_ENTITLEMENT = exports.METERED_DEFAULT_FALLBACK_ENTITLEMENT = exports.NUMERIC_DEFAULT_FALLBACK_ENTITLEMENT = exports.BOOLEAN_DEFAULT_FALLBACK_ENTITLEMENT = void 0;
const sdk_1 = require("@stigg/api-client-js/src/generated/sdk");
__exportStar(require("@stigg/api-client-js/src/generated/sdk"), exports);
/** @internal */
exports.BOOLEAN_DEFAULT_FALLBACK_ENTITLEMENT = {
    type: 'BOOLEAN',
    isFallback: true,
    hasAccess: false,
    accessDeniedReason: sdk_1.AccessDeniedReason.Unknown,
};
/** @internal */
exports.NUMERIC_DEFAULT_FALLBACK_ENTITLEMENT = {
    type: 'NUMERIC',
    isFallback: true,
    hasAccess: false,
    isUnlimited: false,
    accessDeniedReason: sdk_1.AccessDeniedReason.Unknown,
};
/** @internal */
exports.METERED_DEFAULT_FALLBACK_ENTITLEMENT = {
    type: 'METERED',
    isFallback: true,
    hasAccess: false,
    isUnlimited: false,
    hasSoftLimit: false,
    currentUsage: 0,
    requestedUsage: 1,
    accessDeniedReason: sdk_1.AccessDeniedReason.Unknown,
};
/** @internal */
exports.ENUM_DEFAULT_FALLBACK_ENTITLEMENT = {
    type: 'ENUM',
    isFallback: true,
    hasAccess: false,
    enumValues: [],
    accessDeniedReason: sdk_1.AccessDeniedReason.Unknown,
};
/** @hidden */
class CustomerModel {
    constructor(state) {
        Object.assign(this, state);
    }
    getActiveSubscriptions() {
        return this.subscriptions.filter((x) => x.status === sdk_1.SubscriptionStatus.Active);
    }
    getActiveTrials() {
        return this.subscriptions.filter((x) => x.status === sdk_1.SubscriptionStatus.InTrial);
    }
    getActivePromotionalEntitlements() {
        return this.promotionalEntitlements.filter((x) => x.status === sdk_1.PromotionalEntitlementStatus.Active);
    }
}
exports.CustomerModel = CustomerModel;
