// @stigg/node-server-sdk/dist/services/cache/RedisSingleExecutionService/RedisSingleExecution.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedisSingleExecutionService = void 0;
const cacheKeysHelpers_1 = require("../../../utils/cacheKeysHelpers");
const lodash_1 = require("lodash");
const ReuseOngoingExecution_1 = require("../../../utils/decorators/ReuseOngoingExecution");
const RedisSingleExecution_utils_1 = require("./RedisSingleExecution.utils");
/**
 * This class manages the execution of operations in a distributed environment (using Redis).
 * It ensures that an operation is executed only once across multiple processes, and handles notifications between processes.
 */
class RedisSingleExecutionService {
    constructor(operationName, environmentPrefix, 
    // This timeout duration should be slightly longer than the total
    // elapsed time of all retry attempts for the API data fetch request.
    notificationTimeoutMs, redisClient, distributedLocks, loggerService, cacheInstrumentation) {
        this.operationName = operationName;
        this.environmentPrefix = environmentPrefix;
        this.notificationTimeoutMs = notificationTimeoutMs;
        this.redisClient = redisClient;
        this.distributedLocks = distributedLocks;
        this.loggerService = loggerService;
        this.cacheInstrumentation = cacheInstrumentation;
        this.pendingNotifications = new Map();
        this.subscribeClient = redisClient.duplicate();
        this.subscribeClient.on('error', (err) => {
            this.loggerService.error('Redis subscribe client error: ', err);
            this.cacheInstrumentation.trackRedisClientError({ error: err, clientName: 'subscribe' });
        });
        this.subscribeClient.on('connect', () => this.loggerService.log('Redis subscribe client connected!'));
        this.subscribeClient.on('close', () => this.loggerService.log('Redis subscribe client disconnected!'));
        this.subscribeClient.on('message', this.onEventNotification.bind(this));
    }
    async executeOnceAcrossMultipleProcesses(context) {
        const { customerId, resourceId, reloadFromApi } = context;
        const notificationContext = await this.subscribeForNotification(context);
        const { notify, cancelNotificationSubscription } = notificationContext;
        const lockKey = (0, cacheKeysHelpers_1.buildLockKey)(this.environmentPrefix, customerId, resourceId, this.operationName);
        try {
            const response = await this.distributedLocks.usingSingleAttempt(lockKey, async () => {
                // At this point, this process has the responsibility to fetch data from
                // the API, cache it, and then notify other clients to continue their execution.
                // no need to listen anymore as this process is the one who will notify later
                await cancelNotificationSubscription();
                // perform the operation
                return reloadFromApi();
            });
            // notify other clients that the data is ready
            await notify();
            return response;
        }
        catch (err) {
            if (await this.distributedLocks.isReadOnlyReplicaError(err)) {
                throw new Error(`Redis is read-only. Cannot ${this.operationName}`);
            }
            if (this.distributedLocks.isLockTakenByAnotherClientError(err)) {
                // At this point, another process is responsible for fetching data from
                // the API and caching it. This process waits until the data fetching and
                // caching is complete, then retrieves the data from the cache.
                return this.waitForNotificationAndLoadFromCache(context, notificationContext);
            }
            await notify(err.message);
            throw err;
        }
    }
    async waitForNotificationAndLoadFromCache({ customerId, resourceId, reloadFromApi, loadFromCache }, { cancelNotificationSubscription, notificationPromise }) {
        try {
            await notificationPromise;
            return loadFromCache();
        }
        catch (err) {
            await cancelNotificationSubscription();
            if (err instanceof RedisSingleExecution_utils_1.TimeoutError) {
                this.loggerService.error(`Timeout while waiting for notification via Redis for ${this.operationName} - refetching from API`, {
                    customerId,
                    resourceId,
                });
                return reloadFromApi();
            }
            throw err;
        }
    }
    async subscribeForNotification(context) {
        const { customerId, resourceId } = context;
        const notificationChannel = (0, lodash_1.compact)([this.environmentPrefix, this.operationName, customerId, resourceId]).join(':');
        const existingCallbackList = this.pendingNotifications.get(notificationChannel);
        let callbackList;
        if (existingCallbackList) {
            callbackList = existingCallbackList;
        }
        else {
            callbackList = [];
            this.pendingNotifications.set(notificationChannel, callbackList);
            await this.subscribeClient.subscribe(notificationChannel);
        }
        const { notificationPromise, cancelNotificationSubscription } = this.createNotificationPromise(notificationChannel, callbackList);
        const notify = async (error) => {
            try {
                const payload = { error };
                this.loggerService.debug(`Publish notification`, { payload });
                await this.redisClient.publish(notificationChannel, JSON.stringify(payload));
            }
            catch (err) {
                this.loggerService.error('Error while publishing notification', err);
            }
        };
        return { notificationPromise, notify, cancelNotificationSubscription };
    }
    createNotificationPromise(notificationChannel, callbackList) {
        let resolveNotificationPromise;
        let timeoutId;
        const notificationPromise = new Promise((resolve, reject) => {
            resolveNotificationPromise = resolve;
            timeoutId = setTimeout(() => reject(new RedisSingleExecution_utils_1.TimeoutError()), this.notificationTimeoutMs);
            callbackList.push((payload) => {
                clearTimeout(timeoutId);
                if (payload.error) {
                    reject(new RedisSingleExecution_utils_1.RemoteExecutionError(payload.error));
                }
                else {
                    resolve();
                }
            });
        });
        const cancelNotificationSubscription = async () => {
            clearTimeout(timeoutId);
            resolveNotificationPromise();
            await this.unsubscribeChannel(notificationChannel);
        };
        // The promise will be awaited at a later stage. If it gets rejected before
        // that, Node.js will issue an UnhandledPromiseRejectionWarning. To prevent
        // this, we create an orphan promise chain.
        notificationPromise.catch(() => {
            /* ignore */
        });
        return {
            notificationPromise,
            cancelNotificationSubscription,
        };
    }
    async unsubscribeChannel(channel) {
        await this.subscribeClient.unsubscribe(channel);
        const list = this.pendingNotifications.get(channel);
        if (!list) {
            return [];
        }
        this.pendingNotifications.delete(channel);
        return list;
    }
    async onEventNotification(channel, message) {
        try {
            const payload = JSON.parse(message);
            this.loggerService.debug(`Received notification`, { payload });
            const list = await this.unsubscribeChannel(channel);
            for (const callback of list) {
                try {
                    callback(payload);
                }
                catch (err) {
                    this.loggerService.error('Error while executing notification callback', err);
                }
            }
        }
        catch (err) {
            this.loggerService.error('Error while handling notification event', err);
        }
    }
    async cleanup() {
        await this.subscribeClient.quit();
        this.pendingNotifications.clear();
    }
}
__decorate([
    (0, ReuseOngoingExecution_1.ReuseOngoingExecution)(({ customerId, resourceId }) => (0, lodash_1.compact)([customerId, resourceId]).join(':'))
], RedisSingleExecutionService.prototype, "executeOnceAcrossMultipleProcesses", null);
exports.RedisSingleExecutionService = RedisSingleExecutionService;
