// @stigg/node-server-sdk/dist/api/apollo/initApolloClient.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@apollo/client/core");
const isomorphic_ws_1 = __importDefault(require("isomorphic-ws"));
const cross_fetch_1 = __importDefault(require("cross-fetch"));
const utilities_1 = require("@apollo/client/utilities");
const configuration_1 = require("../../configuration");
const requestHeaders_1 = require("../../utils/requestHeaders");
const retry_1 = require("@apollo/client/link/retry");
const WebSocketLink_1 = require("./WebSocketLink");
const TimeoutLink_1 = require("./TimeoutLink");
const OperationTimeoutError_1 = require("./OperationTimeoutError");
// eslint-disable-next-line @typescript-eslint/no-var-requires
const packageJson = require('../../../package.json');
const retryIf = (error) => {
    const doNotRetryCodes = [500, 400];
    const errorIsTimeout = error instanceof OperationTimeoutError_1.OperationTimeoutError;
    return !!error && !doNotRetryCodes.includes(error.statusCode) && !errorIsTimeout;
};
function getAuthLink(apiKey, instanceId) {
    const authLink = new core_1.ApolloLink((operation, forward) => {
        operation.setContext({
            headers: (0, requestHeaders_1.buildRequestHeaders)(apiKey, instanceId, operation.operationName),
        });
        return forward(operation);
    });
    return authLink;
}
function getHttpLink(baseUri) {
    return new core_1.HttpLink({
        uri: `${baseUri}/graphql`,
        fetch: cross_fetch_1.default,
    });
}
function getWsLink(wsUrl, onConnected, apiKey, loggerService) {
    return new WebSocketLink_1.WebSocketLink({
        retryAttempts: Infinity,
        isFatalConnectionProblem: () => false,
        connectionAckWaitTimeout: Infinity,
        keepAlive: 10000,
        url: `${wsUrl}/graphql`,
        on: {
            connected: onConnected,
        },
        webSocketImpl: isomorphic_ws_1.default,
        connectionParams: {
            headers: (0, requestHeaders_1.buildRequestHeaders)(apiKey, loggerService.getInstanceId()),
        },
    }, loggerService);
}
const initApolloClient = ({ apiKey, baseUri, wsUrl, onConnected, loggerService, clientName = configuration_1.SDK_NAME, clientVersion = packageJson.version, timeout, }) => {
    const authLink = getAuthLink(apiKey, loggerService.getInstanceId());
    const httpLink = getHttpLink(baseUri);
    const timeoutHttpLink = new TimeoutLink_1.TimeoutLink(timeout).concat(httpLink);
    const retryLink = new retry_1.RetryLink({
        attempts: { max: 5, retryIf },
        delay: { initial: 500, max: 2000, jitter: true },
    });
    const wsLink = getWsLink(wsUrl, onConnected, apiKey, loggerService);
    const splitLink = (0, core_1.split)(({ query }) => {
        const definition = (0, utilities_1.getMainDefinition)(query);
        return definition.kind === 'OperationDefinition' && definition.operation === 'subscription';
    }, wsLink, retryLink.concat(timeoutHttpLink));
    return new core_1.ApolloClient({
        link: authLink.concat(splitLink),
        cache: new core_1.InMemoryCache(),
        name: clientName,
        version: clientVersion,
    });
};
exports.default = initApolloClient;
