// @stigg/node-server-sdk/dist/api/apollo/WebSocketLink.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebSocketLink = void 0;
const core_1 = require("@apollo/client/core");
const graphql_ws_1 = require("graphql-ws");
const graphql_1 = require("graphql");
class WebSocketLink extends core_1.ApolloLink {
    constructor(options, loggerService) {
        super();
        this.loggerService = loggerService;
        this.client = (0, graphql_ws_1.createClient)(Object.assign({}, options));
    }
    request(operation) {
        return new core_1.Observable((sink) => {
            return this.client.subscribe(Object.assign(Object.assign({}, operation), { query: (0, graphql_1.print)(operation.query) }), {
                next: sink.next.bind(sink),
                complete: sink.complete.bind(sink),
                error: (error) => {
                    const errorReason = error.message || error.reason;
                    if (errorReason.includes('ThrottlerException')) {
                        this.loggerService.log(`Websocket disconnected. reason: Throttle error`);
                    }
                    else {
                        // Error can be also a close event.
                        this.loggerService.error(`Websocket disconnected. reason: ${errorReason || 'connection closed'}`);
                    }
                    sink.error(error);
                },
            });
        });
    }
}
exports.WebSocketLink = WebSocketLink;
